[Work in progress] Limelight aim

kushagharahi opened the pull request:
Opening this for @yuming9 to get a feedback loop going. Work in progress.

elizaBarnett committed:
added limelight stuff

elizaBarnett committed:
added subsystem dependencies for command limelight (other dependencies were done in the gyroscope branch)

elizaBarnett committed:
more high goal stuff

Rene-Baquero committed:
need to adjust constants

Rene-Baquero committed:
fixing

Rene-Baquero committed:
attempt at calling a command with a button

nbgray commented on src/main/java/frc/robot/commands/LimelightTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+    // Make this return true when this Command no longer needs to run execute()
+    @Override
+    public boolean isFinished() {
+        if (limelight.getHorizontalTargetOffset() < DEADBAND_DEGREES) {
```
If we adjust when we're > than DEADBAND, and are finished when we're < DEADBAND, if we're ever exactly at DEADBAND we'll be stuck.  Even though it's extremely unlikely to happen, please change this to <= OR change the execute to >=.  Do not do both.

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -119,6 +131,8 @@ public void teleopInit() {
   public void teleopPeriodic() {
     CommandScheduler.getInstance().run();
+    DrivingDeltas calculatedDeltas = vision.targetDelta();
+      controls.getXButton().whenPressed(new LimelightTopTargetCommand(drive, limelight));
```
Move this to the IO class (soon to be renamed Operator Inputs).

nbgray replied:
Also make this a CommandGroup with ShooterCommand so when aiming is finished the power cells get launched.

nbgray commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -3,8 +3,9 @@
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
-public class Drive {
+public class Drive extends SubsystemBase {
```
We do want this change eventually, just not necessarily a part of this feature.

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -49,8 +59,10 @@ public void robotInit() {
+    this.vision = new Vision(limelight,lights);
+    this.limelight = new Limelight();
     this.lightsCommand = new LightsCommand(lights);
+    this.limelightTopTargetCommand = new LimelightTopTargetCommand(drive, limelight);
```
Move this to IO (soon to be Operator Inputs)

nbgray commented on src/main/java/frc/robot/commands/LimelightTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+    private final double DEADBAND_DEGREES = 0;
+    private Drive drive;
+    private Limelight limelight;
+    private double steeringAdjust = 0;
```
This does not need to be a class variable.

nbgray commented on src/main/java/frc/robot/commands/LimelightTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+            }
+        }
+
+        drive.arcadeDrive(0, steeringAdjust);
```
The being outside the if hOffset > DEADBAND check means we'll keep turning after finding the target.  Move it inside the if (the > DEADBAND if).

nbgray commented on src/main/java/frc/robot/commands/LimelightTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class LimelightTopTargetCommand extends CommandBase {
+    private final double KpAIM = 0.005;
+    private final double DEADBAND_DEGREES = 0;
```
A DEADBAND of 0 is no deadband at all.  Set at 0.5 for now.

nbgray commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,121 @@
+
+    private final double KpAIM = 0.005; 
+    private final double KpDISTANCE = 0.025; // larger than kpAim in initial
+    private final double DEADBAND_DEGREES = 0;
```
A DEADBAND of 0 is no deadband at all.  Set to 0.5 for now.  If this is the same deadband we want for the LimelightTopTarget Command, we might want a constants file so we're not having to change the same variable all over the place.

nbgray commented on src/main/java/frc/robot/commands/LimelightTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+    private Limelight limelight;
+    private double steeringAdjust = 0;
+
+    public LimelightTopTargetCommand(Drive drive,Limelight limelight) {
```
This class name doesn't really match its function.  I suggest renaming it to "TurnToTopTarget" or "AimToTopTarget".  If it's the same command as turning to the bottom target, maybe just "TurnToTarget".

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -119,6 +131,8 @@ public void teleopInit() {
   @Override
   public void teleopPeriodic() {
     CommandScheduler.getInstance().run();
+    DrivingDeltas calculatedDeltas = vision.targetDelta();
```
We don't need this here.  It will be done in LightsCommand, and is part of a different feature.

Rene-Baquero committed:
Fixing According to Nathans Comments

Rene-Baquero committed:
Update Controls.java

Rene-Baquero committed:
Merge branch 'master' into feature/limelightAim

Rene-Baquero committed:
Fixing Merge and adding command x button to call TurnToToTargetCommand

Rene-Baquero committed:
Adding  Constants Class

kushagharahi commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Add your docs here.
+ */
+public class Limelight extends SubsystemBase {
```
Can we change this file to just be a helper class for network tables? Not a subsystembase nor have any other functionality?

nbgray replied:
I'm on board with that.

kushagharahi commented on src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -0,0 +1,120 @@
+  public double getTargetArea() {
+    return NetworkTableInstance.getDefault().getTable("limelight").getEntry("ta").getDouble(0.0);
+  }
+}
```
Can we add setting LimeLight pipelines to this helper class as well?

https://docs.limelightvision.io/en/latest/networktables_api.html

nbgray replied:
Added

kushagharahi commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -0,0 +1,20 @@
+ */
+public class Constants {
+
+    public Constants() {
```
You don't need an empty constructor in a class. The compiler will generate one if it doesn't exist.

nbgray replied:
Removed.  Though I don't like depending on the automatic constructor.  Guess it's the c++ developer in me.

kushagharahi commented on src/main/java/frc/robot/commands/TurnToTopTargetCommand.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public void execute() {
+        double steeringAdjust = 0;
+        final double KpAIM = 0.005;
```
This should be a class variable (IMO)

nbgray replied:
Moved

kushagharahi commented on src/main/java/frc/robot/subsystems/vision/Vision.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Add your docs here.
+ */
+public class Vision {
```
Like @nbgray suggested, let's get rid of this file.

nbgray replied:
Removed

kushagharahi commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -0,0 +1,20 @@
+    public Constants() {
+        //Empty Constructer
+    }
+    public static final double TOP_GOAL_DEADBAND = 0.5;
```
I don't think it's valuable to have this in this Constants file. It's very specific to the LimeLight top goal aim command.

Where I find a Constants file valuable is for all the CAN ids in the system.

nbgray replied:
Removed.  Left Constants file though, for future use.

Rene-Baquero committed:
Adjusted aim command to work correctly

nbgray committed:
Cleaned up code, renamed IO to OperatorInputs, changed TeleopCommand call

nbgray committed:
Made Peer Review changes.

Removed vision class
Moved constant to only place it was used
made Limelight just a Network Tables interface
Added setPipeline to Limelight class

nbgray committed:
Changed shooter to a static class called ShooterMath, added Commands to a SequentialCommandGroup.

nbgray committed:
Added addRequirements(drive) to this command group

kushagharahi commented:
I can't approve my own pull request but i like what's going on in here

nbgray commented:
I want to test it one more time now that I changed the way the teleop command is called and made drive a req for the command sequence.

Commit:
tuned some params, done
